### PR TITLE
ar71xx: Add support to DomyWifi DW33D

### PR DESCRIPTION
**Pull request by mumuqz**
Add support to DomyWifi DW33D.
CPU: QCA9558-AT4A @720MHz
Flash size: 16MB + 128MB
Flash Chip No1: Macronix MX25L128 (16MB)
Flash Chip No2: Micron MT29F1G08ABAEA (128MB) 
RAM size: 256MB
RAM Chip: Nanya NT5TU64M16HG-AC x 2 
Switch: QCA8337N-AL3C
WAN: 1x RJ45 Gbe
Ethernet: 4x RJ45 Gbe
USB: Yes 2 x 2.0
Wireless No1: SoC-integrated: QCA9558-AT4A w/ 3x3 MIMO for 2.4GHz 802.11b/g/n
Wireless No2: Mini PCI-e: QCA9880-BR4A w/ 3x3 MIMO for 5GHz 802.11a/n/ac
Signed-off-by: Jing Lin <mumuqz@163.com>

### CODE REVIEW

File: target/linux/ar71xx/base-files/etc/diag.sh [hunk 1]
```diff
@@ -8,6 +8,9 @@ get_status_led() {
 	local board=$(ar71xx_board_name)
 
 	case $board in
+	dw33d)
+		status_led="domy:blue:status"
+		;;
```
Comment (blogic) [inline target/linux/ar71xx/base-files/etc/diag.sh]: please follow alphabetical ordering

File: target/linux/ar71xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 2]
```diff
@@ -47,6 +47,10 @@ board=$(ar71xx_board_name)
 case "$FIRMWARE" in
 "ath10k/cal-pci-0000:00:00.0.bin")
 	case $board in
+	dw33d)
+		ath10kcal_extract "art" 20480 2116
+		ath10kcal_patch_mac $(mtd_get_mac_binary art 18)
+		;;
```
Comment (blogic) [inline target/linux/ar71xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: please follow alphabetical ordering

File: target/linux/ar71xx/base-files/lib/upgrade/platform.sh [hunk 3]
```diff
@@ -247,6 +247,7 @@ platform_check_image() {
 
 		return 0
 		;;
+	dw33d|\
 	alfa-ap96|\
```
Comment (blogic) [inline target/linux/ar71xx/base-files/lib/upgrade/platform.sh]: please follow alphabetical ordering

File: target/linux/ar71xx/config-4.4 [hunk 4]
```diff
@@ -32,6 +32,7 @@ CONFIG_ATH79_DEV_NFC=y
 CONFIG_ATH79_DEV_SPI=y
 CONFIG_ATH79_DEV_USB=y
 CONFIG_ATH79_DEV_WMAC=y
+CONFIG_ATH79_MACH_DOMYWIFI_DW33D=y
 CONFIG_ATH79_MACH_ALFA_AP120C=y
```
Comment (blogic) [inline target/linux/ar71xx/config-4.4]: please follow alphabetical ordering

File: target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt [hunk 5]
```diff
@@ -1,3 +1,19 @@
...
+        select ATH79_DEV_M25P80
+        select ATH79_DEV_NFC
+        select ATH79_DEV_WMAC
+        select ATH79_DEV_USB
+        select MTD_NAND
+	select MTD_NAND_ECC
+	select MTD_NAND_AR934X
+	select MTD_NAND_AR934X_HW_ECC
+
 config ATH79_MACH_ALFA_AP120C
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt]: please follow alphabetical ordering

File: target/linux/ar71xx/files/arch/mips/ath79/Makefile [hunk 6]
```diff
@@ -38,6 +38,7 @@ obj-$(CONFIG_ATH79_ROUTERBOOT)		+= routerboot.o
 #
 # Machines
 #
+obj-$(CONFIG_ATH79_MACH_DOMYWIFI_DW33D)		+= mach-domywifi-dw33d.o
 obj-$(CONFIG_ATH79_MACH_ALFA_AP120C)		+= mach-alfa-ap120c.o
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/Makefile]: please follow alphabetical ordering

File: target/linux/ar71xx/files/arch/mips/ath79/mach-domywifi-dw33d.c [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+#define DW33D_MAC0_OFFSET		0
+#define DW33D_MAC1_OFFSET		6
+#define DW33D_WMAC_OFFSET		12
+#define DW33D_WMAC_CALDATA_OFFSET	0x1000
+#define DW33D_PCIE_CALDATA_OFFSET	0x5000
+
+static struct gpio_led dw33d_leds_gpio[] __initdata = {
+	{
+		.name		= "domy:blue:status",
+		.gpio		= DW33D_GPIO_LED_STATUS,
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-domywifi-dw33d.c]: leds should use the boardname. please use dw33d:blue:status. same for all of the leds

File: target/linux/ar71xx/files/arch/mips/ath79/machtypes.h [hunk 8]
```diff
@@ -17,6 +17,7 @@
 enum ath79_mach_type {
 	ATH79_MACH_GENERIC_OF = -1,	/* Device tree board */
 	ATH79_MACH_GENERIC = 0,
+	ATH79_MACH_DOMYWIFI_DW33D,		/* DomyWifi DW33D */
 	ATH79_MACH_ALFA_AP120C,			/* ALFA Network AP120C board */
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/machtypes.h]: please follow alphabetical ordering

File: target/linux/ar71xx/image/legacy-devices.mk [hunk 9]
```diff
@@ -108,6 +108,12 @@ define LegacyDevice/AP135
 endef
 LEGACY_DEVICES += AP135
 
+define LegacyDevice/DW33D
+  DEVICE_TITLE := DomyWifi DW33D
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-usb-storage kmod-ledtrig-usbdev kmod-ath10k
+endef
+LEGACY_DEVICES += DW33D
+
 define LegacyDevice/AP136_010
```
Comment (blogic) [inline target/linux/ar71xx/image/legacy-devices.mk]: please follow alphabetical ordering

File: target/linux/ar71xx/image/legacy.mk [hunk 10]
```diff
@@ -240,6 +240,7 @@ ap121_mtdlayout_8M=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,6144k(rootfs
 ap121_mtdlayout_16M=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,10944k(rootfs),4992k(kernel),64k(nvram),64k(art)ro,15936k@0x50000(firmware)
 ap132_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,1408k(kernel),6400k(rootfs),64k(art)ro,7808k@0x50000(firmware)
 ap135_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,14528k(rootfs),1472k(kernel),64k(art)ro,16000k@0x50000(firmware)
+dw33d_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,14528k(rootfs),1472k(kernel),64k(art)ro,16000k@0x50000(firmware);ar934x-nfc:96m(rootfs_data),32m(backup)ro
 ap136_mtdlayout=mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,6336k(rootfs),1408k(kernel),64k(mib0),64k(art)ro,7744k@0x50000(firmware)
```
Comment (blogic) [inline target/linux/ar71xx/image/legacy.mk]: please follow alphabetical ordering

File: target/linux/ar71xx/image/legacy.mk [hunk 11]
```diff
@@ -937,6 +938,7 @@ $(eval $(call SingleProfile,AthLzma,64k,AP121_16M,ap121-16M,AP121,ttyATH0,115200
 $(eval $(call SingleProfile,AthLzma,64k,AP121MINI,ap121-mini,AP121-MINI,ttyATH0,115200,$$(ap121_mtdlayout_4M),RKuImage))
 $(eval $(call SingleProfile,AthLzma,64k,AP132,ap132,AP132,ttyS0,115200,$$(ap132_mtdlayout),KRuImage))
 $(eval $(call SingleProfile,AthLzma,64k,AP135,ap135-020,AP135-020,ttyS0,115200,$$(ap135_mtdlayout),RKuImage))
+$(eval $(call SingleProfile,AthLzma,64k,DW33D,domywifi-dw33d,DW33D,ttyS0,115200,$$(dw33d_mtdlayout),RKuImage))
 $(eval $(call SingleProfile,AthLzma,64k,AP136_010,ap136-010,AP136-010,ttyS0,115200,$$(ap136_mtdlayout),RKuImage))
```
Comment (blogic) [inline target/linux/ar71xx/image/legacy.mk]: please follow alphabetical ordering

File: target/linux/ar71xx/base-files/etc/board.d/01_leds [hunk 12]
```diff
@@ -11,6 +11,12 @@ board_config_update
 board=$(ar71xx_board_name)
 
 case "$board" in
+dw33d)
+	ucidef_set_led_usbdev "mmc" "MMC" "domy:blue:mmc" "1-1"
+	ucidef_set_led_usbdev "usb" "USB" "domy:blue:usb" "2-1"
+	ucidef_set_led_netdev "internet" "INTERNET" "domy:blue:internet" "eth0" "rx"
+	ucidef_set_led_wlan "wlan2g" "WLAN-2.4G" "domy:blue:wlan-2g" "phy1tpt"
+	;;
 airgateway|\
```
Comment (blogic) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: please follow alphabetical ordering

File: target/linux/ar71xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata (hunk 2 above)
Comment (mumuqz) [inline target/linux/ar71xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: Its fixed.

File: target/linux/ar71xx/base-files/lib/upgrade/platform.sh (hunk 3 above)
Comment (mumuqz) [inline target/linux/ar71xx/base-files/lib/upgrade/platform.sh]: Its fixed.

File: target/linux/ar71xx/config-4.4 (hunk 4 above)
Comment (mumuqz) [inline target/linux/ar71xx/config-4.4]: Its fixed.

File: target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt (hunk 5 above)
Comment (mumuqz) [inline target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt]: Its fixed.

File: target/linux/ar71xx/files/arch/mips/ath79/mach-domywifi-dw33d.c (hunk 7 above)
Comment (mumuqz) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-domywifi-dw33d.c]: Its fixed.

File: target/linux/ar71xx/files/arch/mips/ath79/machtypes.h (hunk 8 above)
Comment (mumuqz) [inline target/linux/ar71xx/files/arch/mips/ath79/machtypes.h]: Its fixed.

File: target/linux/ar71xx/image/legacy-devices.mk (hunk 9 above)
Comment (mumuqz) [inline target/linux/ar71xx/image/legacy-devices.mk]: Its fixed.

File: target/linux/ar71xx/image/legacy.mk (hunk 10 above)
Comment (mumuqz) [inline target/linux/ar71xx/image/legacy.mk]: Its fixed.

File: target/linux/ar71xx/image/legacy.mk (hunk 11 above)
Comment (mumuqz) [inline target/linux/ar71xx/image/legacy.mk]: Its fixed.

File: target/linux/ar71xx/base-files/etc/board.d/01_leds (hunk 12 above)
Comment (mumuqz) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: Its fixed.

**Comment by mumuqz** [comment]
@blogic all fixed.

File: target/linux/ar71xx/image/legacy-devices.mk [hunk 13]
```diff
@@ -427,6 +427,12 @@ define LegacyDevice/dLAN_pro_1200_ac
 endef
 LEGACY_DEVICES += dLAN_pro_1200_ac
 
+define LegacyDevice/DW33D
+  DEVICE_TITLE := DomyWifi DW33D
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-usb-storage kmod-ledtrig-usbdev kmod-ath10k
+endef
+LEGACY_DEVICES += DW33D
```
Comment (mkresin) [inline target/linux/ar71xx/image/legacy-devices.mk]: Sorry for being late with the review. But please don't use the legacy build code for new devices. Please use the "new" image build code as it was done with https://github.com/lede-project/source/commit/5d04dcedb4fb773404dcb0625786eb7706284298 for another ar71xx board.

Comment (mumuqz) [inline target/linux/ar71xx/image/legacy-devices.mk]: @mkresin It's fixed.  But build the image size is only 4.7M.  
Error message：The uploaded image file does not contain a supported format.

Comment (mkresin) [inline target/linux/ar71xx/image/legacy-devices.mk]: > It's fixed. But build the image size is only 4.7M.

4.7 MByte sounds pretty normal to me. What was the image size using the old build code? What is the problem with an image of only 4.7 MByte?

>  Error message：The uploaded image file does not contain a supported format. 

Is this error message new or do you had the same error message using the old build code? Where does this error message come from? My first guess is that it's a message from the update webpage of the stock firmware. In that case, you might need to add a factory image to the build code (one with a special header or similar).

Comment (mumuqz) [inline target/linux/ar71xx/image/legacy-devices.mk]: Using the old build code image size is 16 MByte.
Using the old build code to build the image without this error message.
It's a message from the update webpage of the legacy firmware.

Comment (mumuqz) [inline target/linux/ar71xx/image/legacy-devices.mk]: Use “mtd -r wirte lede-ar71xx-generic-dw33d-squashfs-sysupgrade.bin firmware" command successfully, but reboot error message：

[   12.540378] VFS: Cannot open root device "(null)" or unknown-block(31,2): error -5
[   12.548048] Please append a correct "root=" boot option; here are the available partitions:
[   12.597613] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(31,2)
[   12.611010] Rebooting in 1 seconds..

Comment (mkresin) [inline target/linux/ar71xx/image/legacy-devices.mk]: Well, it wasn't meant that you only have to copy/paste **the example** I pointed you to.

The dw33d seams to use an RKuImage (rootfs + lzma compressed kernel with uImage header). The following should work:

```Makefile
define Device/dw33d
  DEVICE_TITLE := DomyWifi DW33D
  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-usb-storage kmod-ledtrig-usbdev kmod-ath10k
  BOARDNAME = DW33D
  IMAGE_SIZE = 16000k
  CONSOLE = ttyS0,115200
  MTDPARTS = spi0.0:256k(u-boot)ro,64k(u-boot-env)ro,14528k(rootfs),1472k(kernel),64k(art)ro,16000k@0x50000(firmware);ar934x-nfc:96m(rootfs_data),32m(backup)ro
  IMAGE/sysupgrade.bin = append-rootfs | pad-rootfs | pad-to 14528k | append-kernel | check-size $$$$(IMAGE_SIZE)
endef
TARGET_DEVICES += dw33d
```

File: target/linux/ar71xx/base-files/etc/diag.sh [hunk 14]
```diff
@@ -126,6 +126,9 @@ get_status_led() {
 	oolite)
 		status_led="$board:red:system"
 		;;
+	dw33d)
+		status_led="dw33d:blue:status"
+		;;
```
Comment (mkresin) [inline target/linux/ar71xx/base-files/etc/diag.sh]: Please use:

```
status_led="$board:blue:status"
```

This change allows to merge multiple identical status led definitions at a later point.

File: target/linux/ar71xx/image/legacy-devices.mk (hunk 13 above)
Comment (mumuqz) [inline target/linux/ar71xx/image/legacy-devices.mk]: @mkresin Its fixed, test passes. Thank you very much!

**Comment by ignisf** [comment]
@mumuqz your commit message is still missing the "Signed-off-by: " line.

File: target/linux/ar71xx/base-files/etc/board.d/01_leds [hunk 15]
```diff
@@ -138,6 +138,12 @@ dragino2)
 	ucidef_set_led_netdev "lan" "LAN" "$board:red:lan" "eth0"
 	ucidef_set_led_netdev "wan" "WAN" "$board:red:wan" "eth1"
 	;;
+dw33d)
+	ucidef_set_led_usbdev "mmc" "MMC" "dw33d:blue:mmc" "1-1"
+	ucidef_set_led_usbdev "usb" "USB" "dw33d:blue:usb" "2-1"
+	ucidef_set_led_netdev "internet" "INTERNET" "dw33d:blue:internet" "eth0" "rx"
+	ucidef_set_led_wlan "wlan2g" "WLAN-2.4G" "dw33d:blue:wlan-2g" "phy1tpt"
+	;;
```
Comment (mkresin) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: Sorry for nagging again. Please use $board here as well.

You can drop the "rx" param from ucidef_set_led_netdev. In that case the LED should react to rx, tx and link.

Comment (mumuqz) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: Its fixed.

**Comment by mkresin** [comment]
As said by ignisf, please use your real name in the Signed-off-by:" line. Otherwise looks good to me now.

**Comment by mumuqz** [comment]
@mkresin I have updated my real name.

File: target/linux/ar71xx/files/arch/mips/ath79/Makefile [hunk 16]
```diff
@@ -84,6 +84,7 @@ obj-$(CONFIG_ATH79_MACH_DLAN_PRO_500_WP)	+= mach-dlan-pro-500-wp.o
 obj-$(CONFIG_ATH79_MACH_DR344)			+= mach-dr344.o
 obj-$(CONFIG_ATH79_MACH_DR531)			+= mach-dr531.o
 obj-$(CONFIG_ATH79_MACH_DRAGINO2)		+= mach-dragino2.o
+obj-$(CONFIG_ATH79_MACH_DOMYWIFI_DW33D)         += mach-domywifi-dw33d.o
 obj-$(CONFIG_ATH79_MACH_EAP300V2)		+= mach-eap300v2.o
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/Makefile]: please use tabs here and not spaces. also DO is before DR, please move the line up by one

File: target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt [hunk 17]
```diff
@@ -542,6 +542,22 @@ config ATH79_MACH_DRAGINO2
...
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_M25P80
+	select ATH79_DEV_NFC
+	select ATH79_DEV_WMAC
+	select ATH79_DEV_USB
+	select MTD_NAND
+	select MTD_NAND_ECC
+	select MTD_NAND_AR934X
+	select MTD_NAND_AR934X_HW_ECC
+
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt]: DO is before DR, please move the line up by one

File: target/linux/ar71xx/config-4.4 [hunk 18]
```diff
@@ -78,6 +78,7 @@ CONFIG_ATH79_MACH_DLAN_PRO_500_WP=y
 CONFIG_ATH79_MACH_DR344=y
 CONFIG_ATH79_MACH_DR531=y
 CONFIG_ATH79_MACH_DRAGINO2=y
+CONFIG_ATH79_MACH_DOMYWIFI_DW33D=y
 CONFIG_ATH79_MACH_EAP300V2=y
```
Comment (blogic) [inline target/linux/ar71xx/config-4.4]: DO is before DR, please move the line up by one

File: target/linux/ar71xx/files/arch/mips/ath79/Makefile (hunk 16 above)
Comment (mumuqz) [inline target/linux/ar71xx/files/arch/mips/ath79/Makefile]: @blogic Its fixed.

File: target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt (hunk 17 above)
Comment (mumuqz) [inline target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt]: @blogic Its fixed.

File: target/linux/ar71xx/config-4.4 (hunk 18 above)
Comment (mumuqz) [inline target/linux/ar71xx/config-4.4]: @blogic Its fixed.

**Comment by mumuqz** [comment]
@blogic When is code merged?

**Comment by blogic** [comment]
merged, Thanks !

**Comment by mumuqz** [comment]
@blogic Why closed with unmerged commits ?

**Comment by jow-** [comment]
Pulled into my staging tree: https://git.lede-project.org/7c8abdf

**Comment by jow-** [comment]
@mumuqz - I suppose it was an oversight, probably blogic forgot to push. I staged your PR in my tree now and will merge it to master ASAP.

**Pull request closed without merging**